BaikalDataGenerator/Source/render.cpp

```diff
@@ -0,0 +1,437 @@
+
+        d.x = elem->FloatAttribute("dirx");
+        d.y = elem->FloatAttribute("diry");
+        d.z = elem->FloatAttribute("dirz");
+
+        r.x = elem->FloatAttribute("radx");
+        r.y = elem->FloatAttribute("rady");
+        r.z = elem->FloatAttribute("radz");
+
+        new_light->SetPosition(p);
+        new_light->SetDirection(d);
+        new_light->SetEmittedRadiance(r);
+        m_scene->AttachLight(new_light);
+        elem = elem->NextSiblingElement("light");
+    }
+}
+
+void Render::LoadSppXml(const std::string& file_name)
+{
+    tinyxml2::XMLDocument doc;
+    doc.LoadFile(file_name.c_str());
+    auto root = doc.FirstChildElement("spp_list");
+
+    if (!root)
+    {
+        THROW_EX("Failed to open lights set file.")
+    }
+
+    tinyxml2::XMLElement* elem = root->FirstChildElement("spp");
+
+    m_spp.clear();
+
+    while (elem)
+    {
+        int spp = (int)elem->Int64Attribute("iter_num");
+        m_spp.insert(spp);
+        elem = elem->NextSiblingElement("spp");
+    }
+}
+
+void Render::UpdateCameraSettings(const CameraInfo& cam_state)
+{
+    if (cam_state.aperture != m_camera->GetAperture())
+    {
+        m_camera->SetAperture(cam_state.aperture);
+    }
+
+    if (cam_state.focal_length != m_camera->GetFocalLength())
+    {
+        m_camera->SetFocalLength(cam_state.focal_length);
+    }
+
+    if (cam_state.focus_distance != m_camera->GetFocusDistance())
+    {
+        m_camera->SetFocusDistance(cam_state.focus_distance);
+    }
+
+    auto cur_pos = m_camera->GetPosition();
+    auto at = m_camera->GetForwardVector();
+    auto up = m_camera->GetUpVector();
+
+    if (cur_pos != cam_state.pos ||
+        at != cam_state.at ||
+        up != cam_state.up)
+    {
+        m_camera->LookAt(cam_state.pos, cam_state.at, cam_state.up);
+    }
+}
+
+void Render::SaveOutput(OutputDesc desc,
+                        const std::string& file_dir,
+                        int cam_index,
+                        int spp)
+{
+    OIIO_NAMESPACE_USING;
+
+    std::stringstream ss;
+
+    ss << "cam_" << cam_index << "_"
+        << desc.name << "_spp_" << spp << ".png";
+
+    std::filesystem::path path = file_dir;
+
+    if (!path.has_filename())
+    {
+        THROW_EX("incorrect output path");
+    }
+
+    path.append(ss.str());
+
+    std::vector<RadeonRays::float3> output_data;
+    std::vector<RadeonRays::float3> image_data;
+    auto output = m_renderer->GetOutput(desc.type);
+    auto width = output->width();
+    auto height = output->height();
+
+    assert(output);
+
+    auto buffer = static_cast<Baikal::ClwOutput*>(output)->data();
+    output_data.resize(buffer.GetElementCount());
+    image_data.resize(buffer.GetElementCount());
+
+    output->GetData(output_data.data());
+
+    for (auto y = 0u; y < height; ++y)
+    {
+        for (auto x = 0u; x < width; ++x)
+        {
+            float3 val = output_data[(height - 1 - y) * width + x];
+            val *= (1.f / val.w);
```

Comment:
Let's add the following comment into the code:
"The 4-th pixel component is a count of accumulated samples. It can be different for every pixel in case of adaptive sampling. So, we need to normalize pixel values here".

Reply:
Done